Fix duplicate check with different mac format

fossedihelm opened the pull request:
A macAddress can be specified with different separators:
 - ":" e.g. "02:00:00:00:ff:ff"
 - "-" e.g. "02-00-00-00-ff-ff"
 - "." e.g. "0200.0000.ffff"

Currently, the duplicate mac check works only if the two compared mac address use the same separator.
The duplicate check is made looking at the existence of key of the macMap. The keys of the mapMap are the macAddress string.

With this commit, a `macKey` struct is introduced. The latter stores the normalized string of the macAddress and can be used as unique key and identifier of a macAddress.

<!-- Thanks for sending a pull request!

Before you click the 'Create pull request' make sure that:
- This PR introduces a single feature of fix, just one
- This PR does not leave the master branch broken
- Every commit in this PR has a commit message explaining what do you change,
  why and what is the outcome
- If your change introduces a complex concept or a change to user interaction
  with the project or the application, make sure to document it

If you don't comply with these rules, you waste your energy, time of reviewers
and cause suffering of future generations.
-->

**What this PR does / why we need it**:

**Special notes for your reviewer**:
Fixes https://issues.redhat.com/browse/CNV-30974

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
3. If no release note is required, just write "NONE".
-->

```release-note
Fix duplicate check with different mac address separator
```

maiqueb reviewed (changes requested):
This PR is required but IMO it doesn't do enough since there are other supported types of format for MAC addresses:
- canonical (`-` separator, which your PR adds support for)
- hexstring (`:` separator, the only thing currently accepted by kubemacpool)
- dot notation (`.` separator, in 4 char segments - e.g. `0000.5e00.5301`)
- hexadecimal 

Maybe the correct thing to do is to create a MAC address - `net.HardwareAddr` - using `net.ParseMAC` and then get the string from it (whatever that is) ?

fossedihelm commented:
> Maybe the correct thing to do is to create a MAC address - `net.HardwareAddr` - using `net.ParseMAC` and then get the string from it (whatever that is) ?

@maiqueb That's a great idea! IMHO we can also skip the `err` from ParseMAC function, since it has already been validated. WDYT?

maiqueb commented:
> > Maybe the correct thing to do is to create a MAC address - `net.HardwareAddr` - using `net.ParseMAC` and then get the string from it (whatever that is) ?
> 
> @maiqueb That's a great idea! IMHO we can also skip the `err` from ParseMAC function, since it has already been validated. WDYT?

Fine with me. I mean, there's nothing wrong in being redundant - checking against nil won't take long - but if you're sure that was already validated, all good. I just find it more readable / comforting to see the (ugly) golang pattern of `if err != nil ; return err` . It's up to the maintainer.

maiqueb reviewed (approved):
Looks good !

fossedihelm commented:
> Fine with me. I mean, there's nothing wrong in being redundant - checking against nil won't take long - but if you're sure that was already validated, all good. I just find it more readable / comforting to see the (ugly) golang pattern of `if err != nil ; return err` . It's up to the maintainer.

Yes, I generally agree! In this  specific case, handling the error means to create something like this:
```
func NewMacKey(macAddressDashes string) (macKey, error) {
	macAddress, err := net.ParseMAC(macAddressDashes)
        if err != nil{
            return macKey{}, err
        }
	return macKey{normalizedMacAddress: macAddress.String()}, nil
}
```
This is not a problem at all, but this will not allow us to create inline entry in the macMap.
And the latter is very useful in tests

maiqueb commented:
> > Fine with me. I mean, there's nothing wrong in being redundant - checking against nil won't take long - but if you're sure that was already validated, all good. I just find it more readable / comforting to see the (ugly) golang pattern of `if err != nil ; return err` . It's up to the maintainer.
> 
> Yes, I generally agree! In this specific case, handling the error means to create something like this:
> 
> ```
> func NewMacKey(macAddressDashes string) (macKey, error) {
> 	macAddress, err := net.ParseMAC(macAddressDashes)
>         if err != nil{
>             return macKey{}, err
>         }
> 	return macKey{normalizedMacAddress: macAddress.String()}, nil
> }
> ```
> 
> This is not a problem at all, but this will not allow us to create inline entry in the macMap. And the latter is very useful in tests, in addition to the fact that we should clearly always handle the return error.

Maybe it's best to add a comment to the `NewMacKey` func indicating it's inputs **must** be already validated MAC addresses ? And indicate all it does is parse from N mac address formats to "our" definition of canonical format.

This is just a nit. Consider adding it if someone else asks you to change the code.

fossedihelm commented:
> Maybe it's best to add a comment to the `NewMacKey` func indicating it's inputs **must** be already validated MAC addresses ? And indicate all it does is parse from N mac address formats to "our" definition of canonical format.
> 
> This is just a nit. Consider adding it if someone else asks you to change the code.

I put this just before the ParseMAC func:
```
// we can ignore the error here since the string value has already been validated
```
but yes, I can insert the description above the func! Thanks, will do!

RamLavi commented:
/ok-to-test

RamLavi commented on pkg/pool-manager/pool.go:
```diff
@@ -79,7 +79,26 @@ type macEntry struct {
 }
 
-type macMap map[string]macEntry
+type macKey struct {
```
I think this one deserves a separate file. Please move this to a new file `macKey.go`, and add the appropriate `macKey_tests.go` with unit tests for it (nothing fancy, I think this is enough:
```
var _ = Describe("MacKey", func() {
	DescribeTable("String should return the normalized MAC address",
		func(input string, expected string) {
			macKey := macKey{normalizedMacAddress: input}
			Expect(macKey.String()).To(Equal(expected))
		},
		Entry("colon-separated input", "00:00:5e:00:53:01", "00:00:5e:00:53:01"),
		Entry("hyphen-separated input", "00-00-5e-00-53-01", "00:00:5e:00:53:01"),
		Entry("period-separated input", "0000.5e00.5301", "00:00:5e:00:53:01"),
	)
})
```
)

fossedihelm replied:
Nice idea! Thanks! Done

RamLavi commented on pkg/pool-manager/virtualmachine_pool.go:
```diff
@@ -431,7 +431,7 @@ func (p *PoolManager) revertAllocationOnVm(vmName string, allocations map[string
 	log.V(1).Info("Revert vm allocation", "vmName", vmName, "allocations", allocations)
 	for _, macAddress := range allocations {
-		delete(p.macPoolMap, macAddress)
+		p.macPoolMap.removeMacEntry(macAddress)
```
ditto, I would move to the preliminary commit

RamLavi reviewed (changes requested):
Hey @fossedihelm, thanks for your contribution!
Please see comments added.

RamLavi commented:
/hold

let's merge only after stable branches are introduced to KMP

fossedihelm commented:
Thanks @RamLavi for the review!
I am working on the relevant failing e2e tests. Thank you!

fossedihelm commented on pkg/pool-manager/macentry_test.go:
```diff
@@ -21,33 +21,33 @@ var _ = Describe("mac-entry", func() {
 		newTimestamp := now.Add(time.Second)
 		BeforeEach(func() {
```
Done!

fossedihelm commented on pkg/pool-manager/pod_pool.go:
```diff
@@ -304,7 +304,7 @@ func (p *PoolManager) revertAllocationOnPod(podFullName string, allocations map[
 
 	log.V(1).Info("Revert vm allocation", "podFullName", podFullName, "allocations", allocations)
 	for _, macAddress := range allocations {
-		delete(p.macPoolMap, macAddress)
```
Done!

fossedihelm commented:
[Changes](https://github.com/k8snetworkplumbingwg/kubemacpool/compare/2479b07cc20aa2c5f3347e5fc9d413852b6f2510..447fd6637834eb509bf739eb7f398a004641f8f5):

- Addressed comments
- Fixed failing e2e test:
--  there were a mixed comparison between `iface.MacAddress` and `macKey.String()`
-- the first one was upper case, the latter in lower case.
- Added `MarshalJSON() ([]byte, error)`  to macMap struct to allow logger to correctly print the map

RamLavi commented on pkg/pool-manager/macpoolmap.go:
```diff
@@ -98,7 +98,7 @@ func (m *macMap) alignMacEntryAccordingToVmInterface(macAddress, instanceFullNam
 	macEntry, _ := m.findByMacAddress(macAddress)
 	for _, iface := range vmInterfaces {
-		if iface.MacAddress == macAddress {
+		if NewMacKey(iface.MacAddress).normalizedMacAddress == macAddress {
```
can we use the `String` method?
```suggestion
		if NewMacKey(iface.MacAddress).String() == macAddress {
```

I think it works, see [here](https://go.dev/play/p/c0PoTSrbmF1)

RamLavi commented on pkg/pool-manager/mackey.go:
```diff
@@ -0,0 +1,22 @@
+	return macKey{normalizedMacAddress: hwMacAddress.String()}
+}
+
+func (m *macKey) String() string {
```
Since we are not changing the value of the object, there is no need to use a pointer to the object
```suggestion
func (m macKey) String() string {
```

This has an added value of being able to call the `String()` right after creating the object: `NewMacKey(MacAddress).String()`

RamLavi reviewed (changes requested):
thanks for the changes! 
Big like for the Marshal function added.

Added one more comment.

RamLavi commented on pkg/pool-manager/virtualmachine_pool.go:
```diff
@@ -528,7 +528,7 @@ func (p *PoolManager) recoverVmFromCluster(macAddress string) (*kubevirt.Virtual
 		}
 
-		if iface.MacAddress != "" && iface.MacAddress == macAddress {
+		if iface.MacAddress != "" && NewMacKey(iface.MacAddress).normalizedMacAddress == macAddress {
```
```suggestion
		if iface.MacAddress != "" && NewMacKey(iface.MacAddress).String() == macAddress {
```

RamLavi reviewed (changes requested):
last one :)

RamLavi reviewed (approved):
/lgtm
/approve

@fossedihelm does this change need to be backported? 
If not, then I'll remove the old after the CVE storm passes and I add the appropriate stable branches...

RamLavi commented:
/hold cancel

RamLavi commented:
/cherry-pick release-0.42
